Fix --set flag

joe-elliott opened the pull request:
**Description:** 
Re-adds the --set parameter in a way that works with `{}` being used in configs.

**Link to tracking Issue:** <Issue number if applicable>
Fixes #1907 

**Testing:**
The original tests were added along with the addition of the following:

```
extensions:
  health_check: {}
```

This failed the previous tests, but now passes.

**Documentation:**

tigrannajaryan commented:
@pavolloffay can you review please?

pavolloffay commented:
Here is the revert https://github.com/open-telemetry/opentelemetry-collector/pull/1905 and the original PR https://github.com/open-telemetry/opentelemetry-collector/pull/1640/files

pavolloffay commented:
@joe-elliott could you highlight what fixed the issue?

joe-elliott commented:
@pavolloffay 
Previously it was iterating through `.AllKeys()` in the config file viper and calling `.Set()` on each one to force them into viper's config "override".  This allowed the `--set` values to nicely merge with the config values.  However, `.AllKeys()` just failed to return a key with an empty object like :  `healthcheck: {}` and so it was never set.

By iterating through all keys and only calling `.Set` on the root keys we are preserving the entire viper config object faithfully.
https://github.com/open-telemetry/opentelemetry-collector/pull/2162/files#diff-2a9451c3c8352620ac89fc278d2e47f08ce3fa76d427cb1694ff8119003caa9fR68-R77

I also attempted other changes like `.MergeConfig()` and can confirm they will not work for this.

joe-elliott commented:
To continue this conversation.  If your file looks like this:

```
a: {}
b:
  c: 2
  d:
```

If you load this with viper and run `AllSettings()` you will get the following structure back:
```
b:
  c: 2
```

If you run `AllKeys()` and log each one you get the following keys:
```
b.c
b.d
```

Viper does not enumerate keys that are empty objects but it will enumerate nil keys.  By calling `Get()` on the root keys only we are able to faithfully reconstruct the original config in the viper overrides so it will nicely merge with the `--set` params.

@pavolloffay thoughts?

joe-elliott commented:
@jpkrohling do you mind taking a look?

tigrannajaryan commented:
> @pavolloffay
> Previously it was iterating through `.AllKeys()` in the config file viper and calling `.Set()` on each one to force them into viper's config "override". This allowed the `--set` values to nicely merge with the config values. However, `.AllKeys()` just failed to return a key with an empty object like : `healthcheck: {}` and so it was never set.
> 
> By iterating through all keys and only calling `.Set` on the root keys we are preserving the entire viper config object faithfully.
> https://github.com/open-telemetry/opentelemetry-collector/pull/2162/files#diff-2a9451c3c8352620ac89fc278d2e47f08ce3fa76d427cb1694ff8119003caa9fR68-R77
> 
> I also attempted other changes like `.MergeConfig()` and can confirm they will not work for this.



> To continue this conversation. If your file looks like this:
> 
> ```
> a: {}
> b:
>   c: 2
>   d:
> ```
> 
> If you load this with viper and run `AllSettings()` you will get the following structure back:
> 
> ```
> b:
>   c: 2
> ```
> 
> If you run `AllKeys()` and log each one you get the following keys:
> 
> ```
> b.c
> b.d
> ```
> 
> Viper does not enumerate keys that are empty objects but it will enumerate nil keys. By calling `Get()` on the root keys only we are able to faithfully reconstruct the original config in the viper overrides so it will nicely merge with the `--set` params.
> 
> @pavolloffay thoughts?

@joe-elliott these ^^^^ are very nice explanations. Would you be able to add them as comments to the relevant parts of the code? Perhaps drop the "previously" part and keep the part that describes how it works now and most importantly why it works correctly.

jpkrohling commented on service/service_test.go:
```diff
@@ -137,7 +144,7 @@ func TestApplication_StartAsGoRoutine(t *testing.T) {
 	params := Parameters{
 		ApplicationStartInfo: componenttest.TestApplicationStartInfo(),
-		ConfigFactory: func(v *viper.Viper, factories component.Factories) (*configmodels.Config, error) {
+		ConfigFactory: func(v *viper.Viper, command *cobra.Command, factories component.Factories) (*configmodels.Config, error) {
```
```suggestion
		ConfigFactory: func(_ *viper.Viper, _ *cobra.Command, factories component.Factories) (*configmodels.Config, error) {
```

You don't seem to require `v` nor `command` there.

jpkrohling commented on service/service_test.go:
```diff
@@ -412,7 +419,7 @@ func createExampleApplication(t *testing.T) *Application {
 	app, err := New(Parameters{
 		Factories: factories,
-		ConfigFactory: func(v *viper.Viper, factories component.Factories) (c *configmodels.Config, err error) {
+		ConfigFactory: func(v *viper.Viper, cmd *cobra.Command, factories component.Factories) (c *configmodels.Config, err error) {
```
```suggestion
		ConfigFactory: func(_ *viper.Viper, _ *cobra.Command, factories component.Factories) (c *configmodels.Config, err error) {
```

jpkrohling commented on service/testdata/otelcol-config.yaml:
```diff
@@ -17,7 +17,7 @@ processors:
   batch:
 
 extensions:
-  health_check:
```
From what I understood, it's still OK to have nil here instead of empty. If that's the case, keep this as nil, as a proof that this change didn't break compat with previous versions.

joe-elliott replied:
Yes, nil is fine here as well as empty object.  The problem is that empty object was not working and was not being tested for.  With my changes this file now tests both empty object and nil object:

```

jpkrohling commented on testbed/testbed/otelcol_runner.go:
```diff
@@ -104,7 +105,7 @@ func (ipp *InProcessCollector) Start(args StartParams) error {
 			GitHash:  version.GitHash,
 		},
-		ConfigFactory: func(v *viper.Viper, factories component.Factories) (*configmodels.Config, error) {
+		ConfigFactory: func(v *viper.Viper, cmd *cobra.Command, factories component.Factories) (*configmodels.Config, error) {
```
Same here :-)

jpkrohling commented on service/set_flag.go:
```diff
@@ -0,0 +1,87 @@
+	// https://github.com/spf13/viper/blob/3826be313591f83193f048520482a7b3cf17d506/viper.go#L1709
+
+	// The workaround is to call v.Set(string, interface) on all root properties from the config file
+	// this will correctly preserve the original config and set them up for viper to overlay them
```
@tigrannajaryan I think this is the explanation you asked for, right?

joe-elliott replied:
I have improved the explanation with some more details about why we are specifically extracting and setting the root keys.

tigrannajaryan replied:
Looks good, thank you.

jpkrohling reviewed (approved):
LGTM. There are a couple of minor improvements that could be made, but those can be safely ignored...

joe-elliott commented:
I believe I've addressed all comments and improved the in code comments regarding the behavior of viper and why we are setting the root keys.

I am getting a failed contribtest:

```
--- FAIL: TestNewLogsExporter (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xbde2df]
```
Could this be due to changes I made?  should I dig into this?

tigrannajaryan commented:
> I am getting a failed contribtest:
> 
> ```
> --- FAIL: TestNewLogsExporter (0.00s)
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
> 	panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xbde2df]
> ```
> 
> Could this be due to changes I made? should I dig into this?

Unlikely that your change could cause this. Looks like a bug in TestNewLogsExporter.

tigrannajaryan commented:
@joe-elliott please rebase from latest master.

joe-elliott commented on service/service.go:
```diff
@@ -103,26 +103,37 @@ type Parameters struct {
-func FileLoaderConfigFactory(v *viper.Viper, factories component.Factories) (*configmodels.Config, error) {
+// The ConfigFactory implementation should call AddSetFlagProperties to enable configuration passed via `--set` flag.
+// Viper and command instances are passed from the Application.
+// The factories also belong the Application and are equal to the factories passed via Parameters.
```
This comment was from @pavolloffay 's original PR, but I believe you are correct.  Fixing!

joe-elliott commented:
Rebased but now getting a build error on contrib:
```
# github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsxrayexporter/translator [github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsxrayexporter/translator.test]
translator/segment_test.go:743:6: val1.InitEmpty undefined (type pdata.AttributeValue has no field or method InitEmpty)
translator/segment_test.go:746:6: val2.InitEmpty undefined (type pdata.AttributeValue has no field or method InitEmpty)
```

tigrannajaryan commented:
> Rebased but now getting a build error on contrib:
> 
> ```
> # github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsxrayexporter/translator [github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsxrayexporter/translator.test]
> translator/segment_test.go:743:6: val1.InitEmpty undefined (type pdata.AttributeValue has no field or method InitEmpty)
> translator/segment_test.go:746:6: val2.InitEmpty undefined (type pdata.AttributeValue has no field or method InitEmpty)
> ```

This is expected, ignore it.
